File: src/orchestrator/state.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use std::path::Path;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectState {
    pub slug: String,
    pub config_path: String,
    pub services: BTreeMap<String, ServiceState>,
    pub started_at: DateTime<Utc>,
    #[serde(default)]
    pub docker: BTreeMap<String, DockerState>,
    #[serde(default)]
    pub compose_services: BTreeMap<String, ComposeServiceState>,
    #[serde(default)]
    pub network_name: Option<String>,
    #[serde(default)]
    pub cluster: Option<ClusterState>,
    #[serde(default)]
    pub dashboard: Option<DashboardState>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClusterState {
    pub cluster_name: String,
    pub kubeconfig_path: String,
    pub registry_name: Option<String>,
    pub registry_port: Option<u16>,
    pub deployed_services: BTreeMap<String, ClusterDeployState>,
    #[serde(default)]
    pub installed_addons: BTreeMap<String, AddonState>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClusterDeployState {
    pub image_tag: String,
    pub last_deployed: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddonState {
    pub addon_type: String,
    pub namespace: String,
    pub installed_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServiceState {
    pub pid: u32,
    pub port: Option<u16>,
    pub port_auto: bool,
    #[serde(default)]
    pub protocol: Option<String>,
    #[serde(default)]
    pub phase: Option<String>,
    #[serde(default)]
    pub exit_code: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DockerState {
    pub container_id: String,
    pub container_name: String,
    pub port: Option<u16>,
    pub port_auto: bool,
    #[serde(default)]
    pub protocol: Option<String>,
    pub named_ports: BTreeMap<String, u16>,
    pub init_completed: bool,
    pub init_completed_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ComposeServiceState {
    pub container_id: String,
    pub container_name: String,
    pub port: Option<u16>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DashboardState {
    pub dashboard_port: u16,
    pub grpc_port: u16,
    pub http_port: u16,
}

impl ProjectState {
    pub fn save(&self, state_dir: &Path) -> anyhow::Result<()> {
        std::fs::create_dir_all(state_dir)?;
        let path = state_dir.join("state.json");
        let content = serde_json::to_string_pretty(self)?;
        // Atomic write: write to tmp file then rename
        let tmp_path = state_dir.join("state.json.tmp");
        std::fs::write(&tmp_path, &content)?;
        std::fs::rename(&tmp_path, &path)?;
        Ok(())
    }

    pub fn load(state_dir: &Path) -> Option<Self> {
        let path = state_dir.join("state.json");
        let content = std::fs::read_to_string(&path).ok()?;
        serde_json::from_str(&content).ok()
    }

    pub fn remove(state_dir: &Path) -> anyhow::Result<()> {
        let path = state_dir.join("state.json");
        if path.exists() {
            std::fs::remove_file(&path)?;
        }
        // Try to remove the directory if it's empty
        let _ = std::fs::remove_dir(state_dir);
        Ok(())
    }

    pub fn state_dir_for(project_dir: &Path) -> std::path::PathBuf {
        project_dir.join(".devrig")
    }

    /// Acquire an exclusive file lock on state.json.lock.
    /// Returns the lock file handle (lock released on drop).
    fn lock_state(state_dir: &Path) -> Option<std::fs::File> {
        let lock_path = state_dir.join("state.json.lock");
        let lock_file = std::fs::File::create(&lock_path).ok()?;

        #[cfg(unix)]
        {
            use std::os::unix::io::AsRawFd;
            // SAFETY: fd is valid for the lifetime of lock_file
            unsafe {
                libc::flock(lock_file.as_raw_fd(), libc::LOCK_EX);
            }
        }

        Some(lock_file)
    }

    /// Atomically update a single service's phase in state.json.
    ///
    /// Uses an exclusive file lock (`flock`) to prevent concurrent
    /// read-modify-write races when multiple grace timers fire at once.
    pub fn update_service_phase(state_dir: &Path, service: &str, phase: &str) {
        let _lock = Self::lock_state(state_dir);
        if let Some(mut state) = Self::load(state_dir) {
            if let Some(svc) = state.services.get_mut(service) {
                svc.phase = Some(phase.to_string());
            }
            let _ = state.save(state_dir);
        }
    }

    /// Atomically update a single service's PID in state.json.
    pub fn update_service_pid(state_dir: &Path, service: &str, pid: u32) {
        let _lock = Self::lock_state(state_dir);
        if let Some(mut state) = Self::load(state_dir) {
            if let Some(svc) = state.services.get_mut(service) {
                svc.pid = pid;
            }
            let _ = state.save(state_dir);
        }
    }

    /// Atomically update a service's phase and exit_code in state.json.
    pub fn update_service_exit(
        state_dir: &Path,
        service: &str,
        phase: &str,
        exit_code: Option<i32>,
    ) {
        let _lock = Self::lock_state(state_dir);
        if let Some(mut state) = Self::load(state_dir) {
            if let Some(svc) = state.services.get_mut(service) {
                svc.phase = Some(phase.to_string());
                svc.exit_code = exit_code;
            }
            let _ = state.save(state_dir);
        }
    }

    pub fn reset_init(&mut self, docker_name: &str) -> bool {
        if let Some(state) = self.docker.get_mut(docker_name) {
            state.init_completed = false;
            state.init_completed_at = None;
            true
        } else {
            false
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use tempfile::tempdir;

    fn test_state() -> ProjectState {
        let mut services = BTreeMap::new();
        services.insert(
            "api".to_string(),
            ServiceState {
                pid: 0,
                port: Some(3000),
                port_auto: false,
                protocol: None,
                phase: None,
                exit_code: None,
            },
        );
        ProjectState {
            slug: "test".to_string(),
            config_path: "devrig.toml".to_string(),
            services,
            started_at: Utc::now(),
            docker: BTreeMap::new(),
            compose_services: BTreeMap::new(),
            network_name: None,
            cluster: None,
            dashboard: None,
        }
    }

    #[test]
    fn update_service_pid_persists() {
        let dir = tempdir().unwrap();
        let state_dir = dir.path();

        let state = test_state();
        state.save(state_dir).unwrap();

        // PID starts at 0
        let loaded = ProjectState::load(state_dir).unwrap();
        assert_eq!(loaded.services["api"].pid, 0);

        // Update PID
        ProjectState::update_service_pid(state_dir, "api", 12345);

        // Reload and verify
        let loaded = ProjectState::load(state_dir).unwrap();
        assert_eq!(loaded.services["api"].pid, 12345);
    }
}

File: internal/state/state.go
// Package state manages persistent devrig project state in .devrig/state.json.
package state

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// ProjectState is serialised to .devrig/state.json.
type ProjectState struct {
	Slug            string                         `json:"slug"`
	ConfigPath      string                         `json:"config_path"`
	StartedAt       time.Time                      `json:"started_at"`
	Services        map[string]ServiceState        `json:"services"`
	Docker          map[string]DockerState         `json:"docker"`
	ComposeServices map[string]ComposeServiceState `json:"compose_services"`
	NetworkName     *string                        `json:"network_name,omitempty"`
	Cluster         *ClusterState                  `json:"cluster,omitempty"`
	Dashboard       *DashboardState                `json:"dashboard,omitempty"`
	// PID of the main devrig process (written so `devrig stop` can signal it).
	PID int `json:"pid,omitempty"`
	// PIDStartTimeMs is the devrig process's creation time (ms since epoch),
	// stored so `devrig stop` can confirm the PID hasn't been reused before
	// signalling it.
	PIDStartTimeMs int64 `json:"pid_start_time_ms,omitempty"`
}

type ServiceState struct {
	PID uint32 `json:"pid"`
	// StartTimeMs is the service process's creation time (ms since epoch),
	// recorded alongside PID so a later run can reuse-proof identify and reap an
	// orphan left by a crashed devrig.
	StartTimeMs int64   `json:"start_time_ms,omitempty"`
	Port        *uint16 `json:"port,omitempty"`
	PortAuto    bool    `json:"port_auto"`
	Protocol    *string `json:"protocol,omitempty"`
	Phase       *string `json:"phase,omitempty"`
	ExitCode    *int    `json:"exit_code,omitempty"`
}

type DockerState struct {
	ContainerID     string            `json:"container_id"`
	ContainerName   string            `json:"container_name"`
	Port            *uint16           `json:"port,omitempty"`
	PortAuto        bool              `json:"port_auto"`
	Protocol        *string           `json:"protocol,omitempty"`
	NamedPorts      map[string]uint16 `json:"named_ports"`
	InitCompleted   bool              `json:"init_completed"`
	InitCompletedAt *time.Time        `json:"init_completed_at,omitempty"`
}

type ComposeServiceState struct {
	ContainerID   string  `json:"container_id"`
	ContainerName string  `json:"container_name"`
	Port          *uint16 `json:"port,omitempty"`
}

type ClusterState struct {
	ClusterName      string                        `json:"cluster_name"`
	KubeconfigPath   string                        `json:"kubeconfig_path"`
	RegistryName     *string                       `json:"registry_name,omitempty"`
	RegistryPort     *uint16                       `json:"registry_port,omitempty"`
	DeployedServices map[string]ClusterDeployState `json:"deployed_services"`
	InstalledAddons  map[string]AddonState         `json:"installed_addons"`
	// K3dVersion is the version of the k3d binary that created/last started the
	// cluster, used to detect version skew on reuse (see cluster.Manager.Ensure).
	K3dVersion string `json:"k3d_version,omitempty"`
}

type ClusterDeployState struct {
	ImageTag     string    `json:"image_tag"`
	LastDeployed time.Time `json:"last_deployed"`
}

type AddonState struct {
	AddonType   string    `json:"addon_type"`
	Namespace   string    `json:"namespace"`
	InstalledAt time.Time `json:"installed_at"`
}

type DashboardState struct {
	DashboardPort uint16 `json:"dashboard_port"`
	GRPCPort      uint16 `json:"grpc_port"`
	HTTPPort      uint16 `json:"http_port"`
}

// StateDir returns the .devrig directory for a project (next to devrig.toml).
func StateDir(configPath string) string {
	return filepath.Join(filepath.Dir(configPath), ".devrig")
}

// Save writes state to state.json atomically (tmp → rename).
func (s *ProjectState) Save(stateDir string) error {
	if err := os.MkdirAll(stateDir, 0o755); err != nil {
		return fmt.Errorf("create state dir: %w", err)
	}
	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal state: %w", err)
	}
	tmp := filepath.Join(stateDir, "state.json.tmp")
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return fmt.Errorf("write state tmp: %w", err)
	}
	return os.Rename(tmp, filepath.Join(stateDir, "state.json"))
}

// Load reads state.json from stateDir. Returns nil if it doesn't exist or
// can't be parsed.
func Load(stateDir string) *ProjectState {
	data, err := os.ReadFile(filepath.Join(stateDir, "state.json"))
	if err != nil {
		return nil
	}
	var s ProjectState
	if err := json.Unmarshal(data, &s); err != nil {
		return nil
	}
	return &s
}

// Remove deletes state.json and attempts to remove the state directory if empty.
func Remove(stateDir string) error {
	path := filepath.Join(stateDir, "state.json")
	if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
		return err
	}
	_ = os.Remove(stateDir)
	return nil
}

// mu protects atomic file-level updates below.
var mu sync.Mutex

// UpdateServiceProc writes the PID and process start time for a service into
// state.json using an exclusive file lock.
func UpdateServiceProc(stateDir, service string, pid uint32, startTimeMs int64) {
	withLock(stateDir, func(s *ProjectState) {
		if s.Services == nil {
			s.Services = make(map[string]ServiceState)
		}
		st := s.Services[service]
		st.PID = pid
		st.StartTimeMs = startTimeMs
		s.Services[service] = st
	})
}

// UpdateServicePhase writes the phase field for a service.
func UpdateServicePhase(stateDir, service, phase string) {
	withLock(stateDir, func(s *ProjectState) {
		if s.Services == nil {
			s.Services = make(map[string]ServiceState)
		}
		st := s.Services[service]
		st.Phase = &phase
		s.Services[service] = st
	})
}

// UpdateServiceExit writes the exit code for a service.
func UpdateServiceExit(stateDir, service string, code int) {
	withLock(stateDir, func(s *ProjectState) {
		if s.Services == nil {
			s.Services = make(map[string]ServiceState)
		}
		st := s.Services[service]
		st.ExitCode = &code
		s.Services[service] = st
	})
}

// withLock loads state.json, applies fn, then saves — holding a file lock and
// the in-process mutex for the duration.
func withLock(stateDir string, fn func(*ProjectState)) {
	mu.Lock()
	defer mu.Unlock()

	path := filepath.Join(stateDir, "state.json")
	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0o644)
	if err != nil {
		return
	}
	defer f.Close()

	// Exclusive advisory lock across processes (multiple devrig instances).
	lockFile(f)
	defer unlockFile(f)

	var s ProjectState
	dec := json.NewDecoder(f)
	if err := dec.Decode(&s); err != nil {
		// File may be empty on first write — use zero value.
		s = ProjectState{}
	}

	fn(&s)

	data, err := json.MarshalIndent(&s, "", "  ")
	if err != nil {
		return
	}
	_ = f.Truncate(0)
	_, _ = f.Seek(0, 0)
	_, _ = f.Write(data)
}
